src/qibolab/instruments/rfsoc.py

```diff
@@ -211,43 +307,50 @@ def play(
             qibolab results objects
         """
 
-        if execution_parameters.acquisition_type is AcquisitionType.RAW:
-            raise NotImplementedError("Raw data acquisition is not supported")
-        if execution_parameters.fast_reset:
-            raise NotImplementedError("Fast reset is not supported")
-        # if new value are passed, they are updated in the config obj
-        if execution_parameters.nshots is not None:
-            self.cfg.reps = execution_parameters.nshots
-        if execution_parameters.relaxation_time is not None:
-            self.cfg.repetition_duration = execution_parameters.relaxation_time
+        self.validate_sequence_and_update_cfg(sequence, execution_parameters)
 
         if execution_parameters.acquisition_type is AcquisitionType.DISCRIMINATION:
             average = False
         else:
             average = execution_parameters.averaging_mode is AveragingMode.CYCLIC
 
-        toti, totq = self._execute_pulse_sequence(self.cfg, sequence, qubits, len(sequence.ro_pulses), average)
+        toti, totq = self._execute_pulse_sequence(sequence, qubits, average)
 
         results = {}
-        adcs = np.unique([qubits[p.qubit].feedback.ports[0][1] for p in sequence.ro_pulses])
-        for j in range(len(adcs)):
-            for i, ro_pulse in enumerate(sequence.ro_pulses):
+        adc_chs = np.unique([qubits[p.qubit].feedback.ports[0][1] for p in sequence.ro_pulses])
+
+        for j in range(len(adc_chs)):
+            channel = sequence.ro_pulses[j].qubit
+            for i, ro_pulse in enumerate(sequence.ro_pulses.get_qubit_pulses(channel)):
                 i_pulse = np.array(toti[j][i])
                 q_pulse = np.array(totq[j][i])
 
-                serial = ro_pulse.serial
-
                 if execution_parameters.acquisition_type is AcquisitionType.DISCRIMINATION:
                     discriminated_shots = self.classify_shots(i_pulse, q_pulse, qubits[ro_pulse.qubit])
                     if execution_parameters.averaging_mode is AveragingMode.CYCLIC:
                         discriminated_shots = np.mean(discriminated_shots, axis=0)
                     result = execution_parameters.results_type(discriminated_shots)
                 else:
                     result = execution_parameters.results_type(i_pulse + 1j * q_pulse)
-                results[ro_pulse.qubit] = results[serial] = result
+                results[ro_pulse.qubit] = results[ro_pulse.serial] = result
 
         return results
 
+    def validate_sequence_and_update_cfg(self, sequence: PulseSequence, execution_parameters: ExecutionParameters):
```

Comment:
This function is good enough, but to improve even more you could split in two separate ones.
Indeed, you never use together `sequence` and `execution_parameters`, so (conceptually) you concatenated two functions because in the present instance you use them only one after the other.